### PR TITLE
Leaderboards functions

### PR DESCRIPTION
**Pull request by yancouto**
Added many leaderboard related functions. (They are being used on MarvInc :D)

Added functions:
- init (no longer called automatically)
- findLeaderboard
- findOrCreateLeaderboard
- getLeaderboard(DisplayType/SortMethod/EntryCount/Name)
- uploadLeaderboardScore
- downloadLeaderboardEntries
- activateGameOverlay
- activateGameOverlayToWebPage
- getFriendPersonaName

Callbacks:
- onUserStatsReceived
- onUserStatsStored
- onUserAchievementStored

Other changes:
- Changed onGameOverlayActivatedCallback to make it more similar to the original API
- Removed outdated README stuff. In the next PR (WIP) I'll add proper documentation.

### CODE REVIEW

**Comment by Kazuo256** [comment]
If the PR is not ready yet, it's better to put a "WIP" somewhere in its title ;)

**Comment by yancouto** [comment]
It's done now :P 
I had just noticed I forgot some stuff while I was writing the documentation so I tried to sneak some changes before the review :)

File: luasteam.cxx [hunk 1]
```diff
@@ -178,15 +520,10 @@ void add_friends(lua_State *L) {
...
-        printf("Sucessfully connected to steam!\n");
-    } else {
-        printf("Couldn't connect to steam...\nDo you have Steam turned on?\nIf not running from steam, do you have a correct steam_appid.txt file?\n");
-        lua_pushboolean(L, false);
-        return 1;
-    }
     global_lua_state = L;
-    lua_createtable(L, 0, 4);
+    lua_createtable(L, 0, 0);
+    uint64Metatable_ref = luaL_ref(L, LUA_REGISTRYINDEX);
```
[the remark this answers is not in the document]
Comment (yancouto) [reply luasteam.cxx]: What I disliked about this method (and checkudata) is that they use strings in the global registry, both cluttering the global registry and also making collisions possible. This way both this problems are avoided. It is more verbose, though.

File: luasteam.cxx [hunk 2]
```diff
@@ -15,26 +15,237 @@ extern "C" {
...
+                pushuint64(L, entry.m_hUGC);
+                lua_setfield(L, -2, "UGC");
+                lua_rawseti(L, -2, ++count);
+            } else {
+                // There really is no reason for this to fail, according to the API
+                // But just in case, we ignore the entry
+            }
+        }
+    }
+    lua_pushboolean(L, io_fail);
```
[the remark this answers is not in the document]
Comment (yancouto) [reply luasteam.cxx]: There is a way [here](https://partner.steamgames.com/doc/api/ISteamUtils#GetAPICallFailureReason), but the documentation says "The failure reasons are typically out of your control and tend to not be very important. Just keep retrying your API Call until it works.".

There's some refactoring I've done (for future PRs) that will help in implementing this, so I'll do this later, for all CallResults.

**Comment by yancouto** [comment]
Instead of always using brackets, I added a clang-format rule for putting if statements without brackets on the same line, and I'm using that for small if statements like

```C++
if (condition) return;
```

This way, there's no confusion because of indentation (and also there wouldn't be normally, since clang-format will indent everything properly).

For large single statements I'm still using braces, since it does look clearer.

```C++
if (condition) {
    this_is_a_very_large_statement(with_a_bunch_of_args);
}
```

------

Also, I didn't add the error messages since the API itself says they're kind of useless, but I'll add them later, once I do some refactoring.

**Comment by henriquegemignani** [comment]
> 
> ```c++
> if (condition) return;
> ```
> 
> This way, there's no confusion because of indentation (and also there wouldn't be normally, since clang-format will indent everything properly).
> 

I prefer to have all `return` statements very clearly visible and not obscured by an `if` statement.

File: luasteam.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kazuo256) [reply luasteam.cxx]: > "The failure reasons are typically out of your control and tend to not be very important. Just keep retrying your API Call until it works."

lol gaben

**Pull request merged**